graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphQLInput.java

```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright 2021 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server.graphql;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.Map;
+import java.util.function.Consumer;
+
+import javax.annotation.Nullable;
+
+import org.dataloader.DataLoaderRegistry;
+
+import com.linecorp.armeria.common.Request;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+import graphql.ExecutionInput;
+
+/**
+ * A input of a GraphQL query over HTTP.
+ */
+public final class GraphQLInput implements Request {
```

Comment:
Can we call `GraphQL.execute(ExecutionInput)` right away in `GraphQLService`instead of having this `GraphQLInput` and `GraphQLOutput` layer?
Are there some features that we can only solve using GraphQL decorator?
If there're not, I think we can completely remove
`GraphQLInput`, `GraphQLOutput`, `GraphQLExecutor` and `DecoratingGraphQLExecutorFunction`.

Reply:
I'll find a way to provide this selectively. 🤔

Reply:
How about keeping `GraphQLExecutor` but using it for the following purposes?

```java
sb.service(GraphQLService.builder()
                         //.build() // ----------------------- (1)
                         .build((grapQL, input) -> { // ------ (2)
                             // pre handle
                             ExecutionResult result = grapQL.execute(input);
                             // post handle
                             return ...
                         }));
```

Reply:
Hm, could you tell me what sort of logic we can put in `pre handle` and `post handle`?
If we cannot handle it in the HTTP decorator layer or in GraphQL, we definitely need it.
However, if we can, we might not need that. 🤔

Reply:
The GraphQL documentation states that we should be able to inspect, transform, modify, etc. the request.
In fact, other frameworks also provide this handling. ([graphql-java-spring](https://github.com/graphql-java/graphql-java-spring/blob/master/graphql-java-spring-webmvc/src/main/java/graphql/spring/web/servlet/ExecutionResultHandler.java), [spring-graphql](https://github.com/spring-projects-experimental/spring-graphql/blob/master/spring-graphql-web/src/main/java/org/springframework/graphql/WebInterceptor.java))

I think it's possible with the decorators we provide, but I think it's inefficient because in order to do this processing (e.g. check, transform, etc.) we have to parse it back to the desired object.

See https://graphql.org/learn/serving-over-http/#web-request-pipeline

Reply:
As discussed in Slack, we're going to get rid of it as it can be replaced by [instrumentation](https://www.graphql-java.com/documentation/master/instrumentation/).